Sources/MapboxNavigation/NavigationMapView.swift

```diff
@@ -2116,16 +2116,61 @@ open class NavigationMapView: UIView {
     @objc func didReceiveTap(sender: UITapGestureRecognizer) {
         guard let routes = routes, let tapPoint = sender.point else { return }
         
-        let waypointTest = legSeparatingWaypoints(on: routes, closeTo: tapPoint)
-        if let selected = waypointTest?.first {
-            delegate?.navigationMapView(self, didSelect: selected)
-            return
-        } else if let routes = self.routes(closeTo: tapPoint),
-                  let selectedRoute = routes.first {
-            delegate?.navigationMapView(self, didSelect: selectedRoute)
-        } else if let alternativeRoutes = self.continuousAlternativeRoutes(closeTo: tapPoint),
-                  let selectedRoute = alternativeRoutes.first {
-            delegate?.navigationMapView(self, didSelect: selectedRoute)
+        routeFromTappedPoint(point: tapPoint)  { [weak self] (routeFoundByPoint) in
+            guard !routeFoundByPoint, let self = self else { return }
+            let waypointTest = self.legSeparatingWaypoints(on: routes, closeTo: tapPoint)
+            if let selected = waypointTest?.first {
+                self.delegate?.navigationMapView(self, didSelect: selected)
+            } else if let routes = self.routes(closeTo: tapPoint),
+                      let selectedRoute = routes.first {
+                self.delegate?.navigationMapView(self, didSelect: selectedRoute)
+            } else if let alternativeRoutes = self.continuousAlternativeRoutes(closeTo: tapPoint),
+                      let selectedRoute = alternativeRoutes.first {
+                self.delegate?.navigationMapView(self, didSelect: selectedRoute)
+            }
+        }
+    }
+    
+    func routeFromTappedPoint(point: CGPoint, completion: @escaping (Bool) -> Void) {
+        let group = DispatchGroup()
+        var routeFoundByPoint: Bool = false
+        let layerIds: [String] = [
+            LayerIdentifier.routeDurationAnnotationsLayer,
+            LayerIdentifier.continuousAlternativeRoutesDurationAnnotationsLayer]
+    
+        for layerId in layerIds {
+            group.enter()
+            let options = RenderedQueryOptions(layerIds: [layerId], filter: nil)
+            routeIndexFromMapQuery(with: options, at: point) { [weak self] (routeIndex) in
```

Comment:
The user would tap on what visuallly appears to be the callout bubble. However, the callout bubble is offset from the actual feature data. This call queries for feature data located at the tap point, which would be some distance away from where the feature data is actually located.

For this approach to be more reliable than `routes(closeTo:)`, we’d technically need to figure out the callout bubble’s screen rect. Alternatively, since we’re filtering to just the annotation layers, we could query for all the features in the current viewport, then select the result that’s closest to the tap point.

Reply:
@ShanMa1991 confirmed that these days, `MapboxMap.queryRenderedFeatures(at:options:completion:)` actually _does_ account for label bounds and offsets. That’s a big deal. I think it resolves this concern and probably also some other issues in our backlog.

/ref https://github.com/mapbox/mapbox-navigation-ios/issues/4165#issuecomment-1255461498